Godeps/_workspace/src/github.com/whyrusleeping/go-multistream/lazy.go

```diff
@@ -0,0 +1,129 @@
+package multistream
+
+import (
+	"fmt"
+	"io"
+	"sync"
+)
+
+func NewLazyHandshakeConn(c io.ReadWriteCloser, proto string) io.ReadWriteCloser {
+	return &lazyConn{
+		proto: proto,
+		con:   c,
+	}
+}
+
+type lazyConn struct {
+	rhandshake bool // only accessed by 'Read' should not call read async
+
+	rhlock sync.Mutex
+	rhsync bool //protected by mutex
+	rerr   error
+
+	whandshake bool
+
+	whlock sync.Mutex
+	whsync bool
+	werr   error
+
+	proto string
+	con   io.ReadWriteCloser
+}
+
+func (l *lazyConn) Read(b []byte) (int, error) {
+	if !l.rhandshake {
+		go l.writeHandshake()
+		err := l.readHandshake()
+		if err != nil {
+			return 0, err
+		}
+
+		l.rhandshake = true
+	}
+
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	return l.con.Read(b)
+}
+
+func (l *lazyConn) readHandshake() error {
+	l.rhlock.Lock()
+	defer l.rhlock.Unlock()
+
+	// if we've already done this, exit
+	if l.rhsync {
+		return l.rerr
+	}
+	l.rhsync = true
+
+	// read multistream version
+	tok, err := ReadNextToken(l.con)
+	if err != nil {
+		l.rerr = err
+		return err
+	}
+
+	if tok != ProtocolID {
+		l.rerr = fmt.Errorf("multistream protocol mismatch ( %s != %s )", tok, ProtocolID)
+		return l.rerr
+	}
+
+	// read protocol
+	tok, err = ReadNextToken(l.con)
+	if err != nil {
+		l.rerr = err
+		return err
+	}
```

Comment:
to me this seems like a special case of the muxer: writing two headers back to back. loses generality. it's fine for this use case, but i'd almost wrap this in two lazyConns, one for each protocol ID.